include/deal.II/base/tensor.h

```diff
@@ -2887,34 +2887,29 @@ constexpr inline DEAL_II_ALWAYS_INLINE Tensor<2, 2, Number>
   return return_tensor;
 }
 
-
 template <typename Number>
 constexpr inline DEAL_II_ALWAYS_INLINE Tensor<2, 3, Number>
                                        invert(const Tensor<2, 3, Number> &t)
 {
   Tensor<2, 3, Number> return_tensor;
 
-  return_tensor[0][0] = internal::NumberType<Number>::value(t[1][1] * t[2][2]) -
-                        internal::NumberType<Number>::value(t[1][2] * t[2][1]);
-  return_tensor[0][1] = internal::NumberType<Number>::value(t[0][2] * t[2][1]) -
-                        internal::NumberType<Number>::value(t[0][1] * t[2][2]);
-  return_tensor[0][2] = internal::NumberType<Number>::value(t[0][1] * t[1][2]) -
-                        internal::NumberType<Number>::value(t[0][2] * t[1][1]);
-  return_tensor[1][0] = internal::NumberType<Number>::value(t[1][2] * t[2][0]) -
-                        internal::NumberType<Number>::value(t[1][0] * t[2][2]);
-  return_tensor[1][1] = internal::NumberType<Number>::value(t[0][0] * t[2][2]) -
-                        internal::NumberType<Number>::value(t[0][2] * t[2][0]);
-  return_tensor[1][2] = internal::NumberType<Number>::value(t[0][2] * t[1][0]) -
-                        internal::NumberType<Number>::value(t[0][0] * t[1][2]);
-  return_tensor[2][0] = internal::NumberType<Number>::value(t[1][0] * t[2][1]) -
-                        internal::NumberType<Number>::value(t[1][1] * t[2][0]);
-  return_tensor[2][1] = internal::NumberType<Number>::value(t[0][1] * t[2][0]) -
-                        internal::NumberType<Number>::value(t[0][0] * t[2][1]);
-  return_tensor[2][2] = internal::NumberType<Number>::value(t[0][0] * t[1][1]) -
-                        internal::NumberType<Number>::value(t[0][1] * t[1][0]);
-  const Number inv_det_t = internal::NumberType<Number>::value(
-    1.0 / (t[0][0] * return_tensor[0][0] + t[0][1] * return_tensor[1][0] +
-           t[0][2] * return_tensor[2][0]));
+  const auto value = [](const auto &t) {
```

Comment:
Can we spell out the type here,
```suggestion
  const auto value = [](const Tensor<2, 3, Number> &t) {
```

Reply:
I think the whole point of the code is to cast whatever the argument is (which is not a tensor, but the product of components of the tensor) to `Number`. I don't recall the issue, but believe that it must be because the product of AD types is not always the AD type itself, and we need to cast it back to that AD type. In other words, the function performs something like a `static_cast`.

So I think that the argument must remain an `auto` type for this to work. But perhaps don't call it `value()`, but `to_Number()`?

Reply:
I agree, thank you!

Reply:
If you don't mind let's stick with `value` (the function in question is also called `value`).

Alternatively, `cast`, or something similar that is at most 8 characters long.

Reply:
I think `value()` just doesn't convey what the function does. The name only conveys that with `NumberTraits` prefixed. But it's not worth arguing too much over that :-)